### ISSUE TITLE
My Subscriptions pagination issues

### ISSUE THREAD

**Issue by benyamin-codez** [anxdpanic/plugin.video.youtube#1035]
### Context

* Add-on Version: 7.2.0+beta.1
* Kodi Version: 20.5
* Kodi GUI Language: English

------

### Expected Behavior

Normal pagination of the "My Subscriptions" item.

------

### Current Behavior

Even pages are missing...?!?!
First next page is 3 (should be 2), second next page is 5 (should be 3), third next page is 7 (should be 4),.... 
Not sure if the data is missing or the pages are just mis-numbered.

------

### Additional Information

Fairly confident this started from version 7.2.0+beta.1.

Logs can be produced if necessary, but thought this would be easily reproduced...

------

**Comment by MoojMidge** [comment]
> Not sure if the data is missing or the pages are just mis-numbered.

Neither. Some items have been filtered out so more than one "page" of data has been used to display the listing. The next page to be listed is therefore not necessarily just the current page number + 1.

**Comment by benyamin-codez** [comment]
> > Not sure if the data is missing or the pages are just mis-numbered.
> 
> Neither. Some items have been filtered out so more than one "page" of data has been used to display the listing. The next page to be listed is therefore not necessarily just the current page number + 1.

That's certainly one perspective... 8^D
... and yes, I had _Shorts_ disabled.

That being said, from a UI standpoint, I would have thought the page number would be for the results _displayed_ and not for a break being imposed on the data itself. Given the previous behaviour was to paginate the results and not the data, wouldn't this be considered a regression...?

**Comment by MoojMidge** [comment]
That is not how the YouTube Data API works, and while My Subscriptions doesn't necessarily need to follow the same rules, it is too much work to implement two different methods for pagination to deal with local filtering unless there is a reasonable need to do so.

Think about it a bit more - if you have 50 items per page and a collection of 1000 items in a list, at what index in the list should the 2nd page start? 50 right? 

What if only 25 items out of the first 50 items are used and 25 item out of the next 50 items? At what index in the list should the 2nd page start? 25? 50? Or 75?

Start to see the problem? Note this is just a simplification, the local "page" index consists of 4 parts - total number, number per page, start index, offset from start index. Essentially each page is relative to an absolute fixed index, which is generally, but not necessarily the first page.

It wouldn't be too difficult to just use a sequential page number for display purposes, but the page number can be used for jumping around in UI, with subsequent pages all relative to this new absolute fixed index.

**Comment by benyamin-codez** [comment]
@MoojMidge 

Yeah, the challenges of using a paginated model like the YT Data API is not lost on me. I might be of some value to wrap it in some sort of abstraction layer.

IIRC, this was previously cached - and maybe still is (I haven't checked the code)...? It seems like it no longer is.

However, a cache could also act as an interim data model. We would then just update the cache from the YT Data API on a periodic basis and only force a full refresh by menu option. We could then produce paginated views from the cache rather than directly from the API. This could also include filtered query results produced from the cache.

The challenge with this approach is deciding how many pages of data does one retrieve from the API. It would likely need to be arbitrary. The other problem is paginating past available data so a trigger to refresh with a larger dataset when paginating past the last page would also be needed.

It is perhaps noteworthy, and IIRC, the cache was originally implemented to reduce hits on the YT Data API endpoint and reduce the likelihood of exceeding usage quotas.

I guess it just seems odd to me to be "skipping" pages.
From a technical standpoint I understand why, yet it still seems odd.

**Comment by MoojMidge** [comment]
> I might be of some value to wrap it in some sort of abstraction layer.

It's not really about abstraction, the pagination details are already abstracted to a single numeric value. To reduce the problem down to its base form - how do you encode all the required paginated details into that one numeric value?

You can't really, but if you make some assumptions (i.e. the number of items per page and the total number of items do not change) and infer some information based on the sequencing of the numeric values (rather than just incrementing the numeric values by one), then you can encode enough information into the single numeric value.

> IIRC, this was previously cached - and maybe still is (I haven't checked the code)...? It seems like it no longer is.

Everything is still cached, just at a lower level than it was before. People complained that their subscriptions weren't being updated quickly enough, so now My Subscriptions is essentially a very simple feed reader with feed history, with the historical content of the feeds being stored, and the underlying video and channel data being cached and updated as needed.

> We could then produce paginated views from the cache rather than directly from the API.

This is done automatically, but at the end of the day you can't cache everything, and whenever you query via the API you are going to be limited to a maximum of 50 items per query.

> This could also include filtered query results produced from the cache.

This could be done, but it is really complicated because there can potentially be multiple filtered views all being accessed at the same or different time from different sources.

> I guess it just seems odd to me to be "skipping" pages.
> From a technical standpoint I understand why, yet it still seems odd.

Perhaps, but compared to what YouTube themselves do (i.e. no manual pagination possible), then it does offer a lot more flexibility, and also information about the fact that results are being filtered.

Still open to ideas if you have anything further.

**Comment by benyamin-codez** [comment]
@MoojMidge 
cc: @kedarz

To avoid tickets, perhaps it's worth considering a qualitative approach rather than a quantitative one.

This might be achieved with a _"[Filtered]"_ label or similar, resulting in e.g. _Next Page [Filtered] 2..._ but used with sequential numbering. In this way, a user will see that the results are indeed filtered yet not have "missing" pages. This would likely be sufficient for support too, e.g. when determining if a listing has passed through the filtered code path.

**Comment by MoojMidge** [comment]
> It wouldn't be too difficult to just use a sequential page number for display purposes, but the page number can be used for jumping around in UI, with subsequent pages all relative to this new absolute fixed index.

The problem is still this. How can the sequential numbering be converted to the actual next page? If it is just cosmetic then it is easy enough to format the text label using a new additional query parameter, just didn't see the point.

Edit: Indication that content is filtered added, sequential numbering too much hassle for little benefit.

**Comment by benyamin-codez** [comment]
@MoojMidge 

Thanks for having another look.

> The problem is still this. How can the sequential numbering be converted to the actual next page?

I'm likely too far removed from the current implementation to really comment, but...

Is that a limitation of the view? I would have thought the control would be agnostic as to any page number and rely just on it being _next_, i.e. perform a simple index increment. If we know which is the last item on page X, then the first item on page X+1 should just be the next item in the list.

> Edit: Indication that content is filtered added, sequential numbering too much hassle for little benefit.

That might stop the tickets... Time will tell.

Aesthetically, imho, I don't think "(filtered)" is as appealing as "(Filtered)".

**Comment by MoojMidge** [comment]
There may be a data structure or model that could be used to do this, but I just can't think of one that would work.

<details>


<summary>An explanation which includes some contrived complications to demonstrate the complexity</summary>

---

Assuming that a complete list of videos is available (it isn't, which renders the following problem simplification a bit useless, but it should be enough to kind of explain), lets also assume you have 1000 items in total, 50 per paged API response:

1. 1st page view takes 40 items from the first API response, then 1 item only for the next 10 responses (this will vary, it could be 5 items for one response, 0 items matching the filters, or evan all 50 matching in which case some are discarded).

2. 2nd page view is actually starting from the 12th API response.

3. If you want to jump to the 3rd or any arbitrary page view how do you know where to start from? You have to index based on the next to current page view then add `x * y` to the index, where `x` is the page view offset from the next to current page view and `y is the number of items per API response.

Now assuming that you still have the complete list available (again it isn't actually available until you navigate through the listings) how do you pass the list around from one plugin invocation to another in order to index it based on the last item from the current page view?

The next page item has url query parameters that indicate various parameters related to the API responses, but to store the actual list of videos in the plugin url query parameters is not practical. Storing the list of videos in the cache is possible, but basically means that every single filtered view needs to be stored so that it can be retrieved when the next page in the view is accessed.

If it were possible to only ever have one view this would be simple, but you can have multiple plugin invocations and multiple views. So the next problem is how to match the list (again not actually available) for a view with the plugin invocation so that it can be retrieved from the data cache? This could be done by checking the plugin url for the plugin invocation.

But hang on - the results for the same plugin url can vary depending on the filters that are used.

Ok, we can workaround that by just adding the currently applied filter as a plugin url query parameter (which is functionality that already exists to allow individual widgets to have different filters to tho those set in the addon settings).

But what happens if the filters change? Well now our url query parameters are already set and can't be changed because the next page item has already been created and can no longer be modified.

Ok, maybe we can just check that whether the url parameters have a mismatch compared to the filter settings at the start of the plugin invocation and use the current settings instead (lets just ignore the fact that the filter parameter is intended to override the filter settings, and just assume any mismatch means that the settings should apply).

Ok so now that we have a new set of filters, lets retrieve our list. Oh well now the list is no longer valid for the view. So we are back to the original problem of how to know what items to include in the current view.

Just for the sake of having consecutive next page numbers, these are problems I'd rather just ignore, even though I have thought about what is actually practical to do.
</details>

**Comment by benyamin-codez** [comment]
@MoojMidge 

I had a very brief look over your more recent work starting at `get_my_subscriptions()`.
You've certainly been busy..! Thank you by the way for all your work and for your detailed response(s).

Architecturally, I would probably stick with MVC. Perhaps you already are? I see some elements of it.

If so, do we treat the cache as the model? It seems we treat the YT Data API as the model rather than a data source.

If we were to use the cache as the model: Whatever the tablespace mechanism (and I'm not familiar with what is used - is it `xml.etree.ElementTree`?), all views should be derived from this model via a controller endpoint accessible to the view. The means of updating the data in the model, i.e. the cache, would be a separate, dedicated refresh controller that performs this as a triggered function in isolation to any view or the view controller. Moreover, each view should make its query without any regard as to the staleness of the data. The refresh controller would perform data updates on an as-needed basis, whilst accepting refresh requests from the view controller. These may or may not be blocking requests, i.e. whether the view controller waits for a response or not before returning (presenting) the view to the user. The need to perform a refresh (and the resultant response - whether preliminary, interim or final) would be determined by a comparison with a TTL setting. The refresh query* performed should satisfy the data requirements of all other possible queries.

[*] if not a single query, perhaps then a small subset of queries, those necessary to create a usable index.

I would have thought that via such an architecture, most of the issues you raised could be overcome.

Of note, you mentioned:

> Storing the list of videos in the cache is possible, but basically means that every single filtered view needs to be stored so that it can be retrieved when the next page in the view is accessed.

The alternative here is that you store the list of videos in the cache, but provide a view controller that functions as an API equivalent endpoint (or other functionally equivalent mechanics) that returns any view query in the expected manner.

In terms of the TTL, this could be set from options like _Before/With every request_, or _After every video finishes_, to an arbitrary amount of time - whether seconds, minutes or longer (where the refresh is scheduled, and occurs during idle or is otherwise queued for when the video finishes) - to _Never_, or even the instructional _Manual refresh only_. Whilst I'm sure some users need their feed to be constantly fresh, I don't mind mine being a little stale, especially if it improves my user experience (it does).

By owning the data in the model, with it's finite and known structures, views and resultant pagination could likely be greatly simplified - the queries themselves would certainly not need to be page indexed as is the case with the YT Data API. Only the refresh mechanics would need to be concerned with this, and less so, if we are just grabbing all the relevant data we can.

Anyway, just my 2c and perhaps you're already doing some of this, but if not, maybe this is a roadmap to consider. 8^d

**Comment by MoojMidge** [comment]
> I would have thought that via such an architecture, most of the issues you raised could be overcome.

Not really. Following the MVC pattern the issue is with the controller.

> The alternative here is that you store the list of videos in the cache, but provide a view controller that functions as an API equivalent endpoint (or other functionally equivalent mechanics) that returns any view query in the expected manner.

This is the crux of the original problem - how does a single incrementing numerical input get translated by the controller to produce any arbitrary view? It is irrelevant whether the content of that view is stored or generated, as the issue is a data encoding problem.

YouTube itself uses a page token, which is a protobuff encoded value that contains all the details of the current/next paged view. It is impossible to replicate the same thing using only a single incrementing numerical input.

**Comment by benyamin-codez** [comment]
> This is the crux of the original problem - how does a single incrementing numerical input get translated by the controller to produce any arbitrary view? It is irrelevant whether the content of that view is stored or generated, as the issue is a data encoding problem.

As discussed above, my thinking was that you would have two controllers: one for refreshing data and one for creating views. **_You wouldn't necessarily need to use the same queries for each controller._**

As for generating views, the page index would simply be a function of the result and results per page setting, i.e. the division of the result by page. So when the results per page setting is say 50, a view request for page 1 would return results 1 to 50, and view request for page 2 would return results 51 to 100, i.e. a page index in its simplest form.

I would suggest the view query could be simplified and not contain any filtering parameters. As the view settings are essentially global, the refresh controller, would likely fetch data that is filtered - and that would certainly be appropriate. As such any views generated by the view controller would _agnostically_ return a filtered result.

So whilst the view controller (or a third controller) might return what constitutes a user's preferred filter parameters from the UI, it would not _intentionally_ generate filtered views. Instead the refresh controller would source filtered data from the data source (i.e. the YT Data API) based on these parameters when it is populating the model (i.e. the cache) from which the views are generated.

This is just one method to achieve both data and UI abstraction of the YT Data API in relation to the generation of views. By doing so, you no longer need to encode any data into the page index or make use of YT page tokens, etc. because you would no longer be using the YT Data API to generate the view (it is only used to refresh the cache data), and instead you would make use of your own (simplified, isolated and abstracted) controller to generate any view.

It is noteworthy a solution like this would likely also provide an opportunity to simplify the refresh controller queries also.

Am I making any sense...?

**Comment by MoojMidge** [comment]
Unfortunately you are just handwaving away some very important implementation details.

> As discussed above, my thinking was that you would have two controllers: one for refreshing data and one for creating views. **_You wouldn't necessarily need to use the same queries for each controller._**

These processes aren't tightly coupled. If the videos in a view are known, then data is requested directly.

> I would suggest the view query could be simplified and not contain any filtering parameters. As the view settings are essentially global, the refresh controller, would likely fetch data that is filtered - and that would certainly be appropriate. As such any views generated by the view controller would _agnostically_ return a filtered result.

As mentioned before this cannot just be assumed to be case and would remove the ability to have different views in different widgets, amongst other things. It would also mean that filtering shorts or live videos out generally would also mean that they could not be viewed in the dedicated shorts or live video folders when those were accessed. But regardless, that is not the problem.

> As for generating views, the page index would simply be a function of the result and results per page setting, i.e. the division of the result by page. So when the results per page setting is say 50, a view request for page 1 would return results 1 to 50, and view request for page 2 would return results 51 to 100, i.e. a page index in its simplest form.

This is the problem. How is it possible to know what the next 50 items are if you haven't already retrieved them? Noting that this could take multiple paged requests from the YouTube API to do, because you don't know whether any of the next video items need to be shown until you have already fetched them.

> This is just one method to achieve both data and UI abstraction of the YT Data API in relation to the generation of views.

You can't just abstract away the relationship, at some point there still needs to be an implemented link between what you want to show and what you are requesting.

When you don't know what to show next (the next 50 items), you have to request whatever is available to show and then decide whether to show it or not, so in order to show the next 50 items you may have to request 100 items, or 150, or 200, or 500, or the entire list of whatever is available.

So when you want to show the next 50 items again (i.e. the page you are viewing - one variable), you need to ensure that you identify at what point you previously stopped retrieving items (the current index - a second variable), so that you can start getting the next lot of items from that point (in paged batches of up to 50, but can vary - a third variable) and stop when you have all the items you want (50 in this example - another variable but can be inferred from the first variable) or when the total number of available items have been retrieved (a fourth variable).

You simply cannot encode this data as a single numeric page number.

Think of it this way - if I ask you to open any random book and then flip through the pages, starting from the first, until you had counted exactly 100 instances of the word "the", could you also flip through the pages containing the second hundred instances of the word "the" without checking the first hundred to know where to start?

**Comment by benyamin-codez** [comment]
> Unfortunately you are just handwaving away some very important implementation details.

Most likely. 8^d
Please accept my apologies. 🙏 

>> As such any views generated by the view controller would agnostically return a filtered result.
>
> As mentioned before this cannot just be assumed to be case and would remove the ability to have different views in different widgets, amongst other things. It would also mean that filtering shorts or live videos out generally would also mean that they could not be viewed in the dedicated shorts or live video folders when those were accessed.

You could have more than one model accessed by a common view controller, but a more common method would be to use shared models, where data sets resulting from a single refresh query can service multiple views, and in so doing reduce the required number of models. This also doesn't have to be made generic, it could be specific, for example to _My Subscriptions_ and _Recommendations_.

> This is the problem. How is it possible to know what the next 50 items are if you haven't already retrieved them? Noting that this could take multiple paged requests from the YouTube API to do, because you don't know whether any of the next video items need to be shown until you have already fetched them.

Many pages from the YT API would need to be fetched to initially populate the model. Subsequent refreshes might only query for recently published videos. Once the data is in the model, you can then query the model to determine metadata such as number of records, etc., to use in formulating queries for views.

> Think of it this way - if I ask you to open any random book and then flip through the pages, starting from the first, until you had counted exactly 100 instances of the word "the", could you also flip through the pages containing the second hundred instances of the word "the" without checking the first hundred to know where to start?

Presuming a SQL back end, one can use `LIMIT` and `OFFSET`:

```sql
SELECT relevant_cols FROM my_model [WHERE conditional] [ORDER BY col] LIMIT items_pp OFFSET start_item
```

For example, if returning data for page 3 when using a value of 50 items per page, you can use:

```sql
SELECT col_a, col_b FROM my_model LIMIT 50 OFFSET 100

SELECT col_a, col_b FROM my_model LIMIT 100,50
```

Would that be of use..?
I'm not sure what query mechanics are currently in use...
...but I did see some SQL in `resources/lib/youtube_plugin/kodion/sql_store/storage.py`...

**Comment by MoojMidge** [comment]
> For example, if returning data for page 3 when using a value of 50 items per page, you can use:
> 
> SELECT col_a, col_b FROM my_model LIMIT 50 OFFSET 100
> 
> SELECT col_a, col_b FROM my_model LIMIT 100,50

Where does the value of `100` come from? It is a view offset given by `view_offset = (next_page - 1) * items_per_page`, so that is two pieces of information required, but you are only given one - the `pnext_age`. However we can infer `items_per_page` or otherwise assume it does not change.

Of course this presumes that every row in your table or view model can be used in your paged view, and you can just select them after ordering based on whatever condition is specified.

However you do not know any of this beforehand. You can only do this once you have built up your view model and the data that is used can not be sequentially selected out of a table or view.

In order to get more data into your nicely ordered and relevant view model you need to pick them out of the paged data model. For the purposes of illustration lets say every 20 items out of 50 from the paged data model query response is useful/relevant.

So if current page is `2`,  `next_page` is `3` then `view_offset` is `100`, however `data_offset` would be `view_offset +((50 - num_useful_items_per_page) * (next_page - 1)) = 160`.

So in order to fill your view model up with items you now need to know, as a minimum, `next_page`, `view_offset` (calculated from `next_page`), and `num_useful_items_per_page`.

Again the input to the view model is all we have i.e. `next_page`, so we are missing `num_useful_items_per_page`. Also it is not a constant (20 that was assumed for this example), it will change for every single paged view and also for every single combination of applied filters. It could be 50, 10, 20, 0, or 1 and so for every single `next_page` you will have a varying `data_offset`.

It doesn't matter how much you abstract this, or how many different models you have, at the end of the day there must be a link between the `next_page` and the `data_offset`, because `next_page` is the only input to your view model.

What is therefore done is to calculate the effective `next_page` value based on the current known `data_offset` for the given `current_page`. If `current_page` is `2` and the known current `data_offset` is `160`, then `next_page` is `4`, rather than `3`

**Comment by MoojMidge** [comment]
> Many pages from the YT API would need to be fetched to initially populate the model.

I think maybe this is what you are getting stuck on. For a given data query there may 1000 results. It is simply impractical to fetch all 1000 result to populate the view model on the off chance that item 990 is relevant to the 20th page that is viewed when you are only viewing the first page.

It would mean that the first page will take 20x times as much time to populate, and would also burn through your quota allocation. You would also start filling your cache up which would mean that more relevant results fetched previously, would then get bumped out of the cache, simply to store less relevant results on the off chance they may be eventually be viewed.

Is this an appropriate trade off to get consecutive next page numbers? No

**Comment by MoojMidge** [comment]
> Presuming a SQL back end, one can use `LIMIT` and `OFFSET`:
> 
> SELECT relevant_cols FROM my_model [WHERE conditional] [ORDER BY col] LIMIT items_pp OFFSET start_item

Also this would only be relevant to the book example if every single page in the book was  full text indexed beforehand i.e. fully read. Why would you bother doing that if you were likely to only read the first chapter, and it would take you days before you could even start on the made up task of flipping through the pages to find the second group of 100 "the" words.

**Comment by benyamin-codez** [comment]
🤔

**Comment by ColonelPanic-42** [comment]
I have yet to have a chance to look into this properly (I'd like to contribute more than just being another bug reporter if I can 😂), but I have found that with _any_ sort of filtering enabled that triggers this workflow (filtering shorts/etc, custom filters, etc), videos tend to get skipped between pages.

To try and provide an example, loading page 1 will get 50 videos from the _real_ page 1, filter out say 20 of them, load the real page 2 and get the next 20 suitable videos, which is all well and good. The next page link shows "Next page: 3 (filtered)", which then loads and appears to start from the _real_ page 3, which means the potentially up to 30 other videos from page 2 get skipped. Skimming the comments on this thread so far I was under the impression the intent was to somehow cache the page results and stitch them together so the videos in between don't get skipped, but either I've misunderstood or otherwise missed something said, or it's not working as intended at least on my setup.

I am hoping to get some time this weekend to have a proper look at this. @MoojMidge is there anything in particular you'd like me to do to produce logs etc. if you're still working on this at all, otherwise I'll just see what I can discover, if I can polish the rust off my Python experience and work out how this all works 🤣 

**_EDIT_**: After having a closer read of this thread, I think this might be an intended but undesirable behaviour, according to this bit from https://github.com/anxdpanic/plugin.video.youtube/issues/1035#issuecomment-2600818562 (emphasis mine):
> 1. 1st page view takes 40 items from the first API response, then 1 item only for the next 10 responses (this will vary, it could be 5 items for one response, 0 items matching the filters, or evan all 50 matching **in which case some are discarded**).
>
> 2. 2nd page view is actually **starting from the 12th API response**.

TBH I don't really care on way or another about what number to display at the bottom of the page, it's literally just a marker to say "click here to show what comes after this", and I can definitely appreciate not wanting to bend over backwards to fix what's essentially a display issue. But actually excluding videos between pages because they get discarded is IMO a significant regression and a bug. The previous behaviour might have resulted in pages having less items than the configured page size, but at least you could be sure you didn't miss anything as you paged through; it took me a couple of months to realise I'd missed dozens of videos from my subscriptions because of this and had to disable _all_ filtering just to make sure I don't miss anything again (which is a pain now that more than half of the results are shorts now 😩).

I'm really hoping there's an alternative solution that can resolve this 🤞

**Comment by benyamin-codez** [comment]
@ColonelPanic-42 

Thanks for your post. That's a very interesting observation. I believe we now fetch subscriptions sorted by last active. I read somewhere recently (soz, I have been unable to find the article /post again) that the YT Data API may not return a channel that is below a certain update frequency, i.e. it shouldn't necessarily be read as this channel was updated on this date. This could be a potential cause for certain channels being dropped from the results returned by the YT Data API.

cc: @MoojMidge

**Comment by ColonelPanic-42** [comment]
OK, so I am _quite_ rusty when it comes to Python 😅 and I've had no experience with writing Kodi plugins, so atm I have NFI how to resolve the issue, but I can definitely confirm the code as it currently is appears to discard surplus items between pages.

In `resources/lib/youtube_plugin/youtube/helper/v3.py`, starting from line 651 in the current version, this block of code here:
```python
if items:
    if remaining and remaining < len(items):
        items = items[:remaining]
    filtered_items.extend(items)

remaining = num_original_items - len(filtered_items)
if remaining >= 0:
    if next_page:
        next_page += 1
    elif current_page:
        next_page = current_page + 1
    else:
        next_page = 2
if remaining <= 0:
    break
```
When the number of filtered items returns, and is larger than the remaining items for the page, the list is truncated to the remainder count, and then added to the full collection of filtered items for the page. The remainder is updated to 0, and then the `next_page` is incremented and the loop breaks. But the code never goes back and deals with those items truncated from the `items` collection, and since the page has been incremented the next query reads the following page and skips over the truncated items.

IDK if this is oversimplifying things, but could it be possible to store in the cache a relatively short-lived `offset` value for a page equal to the number of items truncated, and then rather than incrementing the `next_page` when items have been truncated instead have it check the cache if an offset value exists for that page (if it's the initial read for a request; subsequent pages required to fill the items shouldn't read the offset since they need to be sequential) and if so then skip the first items on the return equal to the offset? A positive side effect would also be _slightly less_ discrepancy of the actual vs displayed page count (since the next page counter wouldn't be incrementing as often).

**Comment by ColonelPanic-42** [comment]
OK I made a POC PR https://github.com/anxdpanic/plugin.video.youtube/pull/1143, not sure if it's the best way to go about it given my lack of experience with Kodi dev and understanding of this plugin in general, but it seems to work with my manual testing on the My Subscriptions page so 🤞 . If it's close enough, feel free to tweak or whatever you want/need to fit with the intended implementation, or if I'm missing something significant then hopefully it provides some useful ideas? I tried to consider how it would work with other potentially filtered lists so in theory it should work for other lists as well.

**Comment by MoojMidge** [comment]
TLDR:

> the intent was to somehow cache the page results and stitch them together so the videos in between don't get skipped

Yep.

> it's not working as intended

It is a work in progress to combine various filtering and filler methods for different types of listings as the filtering, filler and list generation methods have changed over the course of this year.

> I think this might be an intended but undesirable behaviour

It is undesireable, and not intended, Just not currently working.

> if this is oversimplifying things, but could it be possible to store in the cache a relatively short-lived `offset` value for a page equal to the number of items truncated

It is oversimplifying things.

The offset was previously being passed around as a plugin url query parameter, this can be done again without needing to specifically cache it. Storing the offset value in the cache means that only one offset value can be exist at a time for the given My Subscriptions plugin path, while it is possible for a number of different My Subscription listings to exist at the same time with the same or different query parameters.

The offset parameter was removed. It needs to be replaced with a different plugin url query parameter that lists the actual videoIds of the next page that are used to fill the current page, so that these videoIds can be excluded from the display of the next page.

This is already done, but I am working on a few other things related to the Recommended videos. Will try to make come commits with the changes in the next day or two for testing.

@ColonelPanic-42 - the issue that you have identified is valid, and is related to a number of different issues, rather than this original issue specifically.

Originally the only filtering that was done by this addon was for whole channels in My Subscriptions, which was fairly simple since it didn't actually impact on how a video listing was put together, it just changed the content used to generate the listing. I'll refer to this as input filtering.

Separate filtering was added for shorts, which was then extended to other types of videos/stream and then further extended to (almost) full custom filtering of any video/stream based on any available metadata.

However this filtering was done after the listings were already generated, which resulted in listings that would vary in size, depending on how many of the input videos/stream were filtered out. I'll refer to this as output filtering.

Output filtering was done as one of the last stages of generating a listing in the `v3.response_to_items` method, and applied to all types of listings in the same way.

Input filtering, on the other hand, was done in an ad-hoc manner and was implemented in different ways for different types of specific listings (My Subscriptions, Recommended, Related, etc.), because the source of the data used to create the listings was different and the methods to generate the listings were therefore also different.

In order to resolve the problem where output filtering would result in listings of varying size, the concept of listing fillers was added, which would allow a filtered listing to be filled with additional content to fully populate a listing.

The end goal is that when output filtering was used, listing fillers would be used to fully complete a listing. Custom fillers were added for some listings, but not all, with the eventual aim to remove all custom fillers and do this automatically for all listings.

Doing this automatically for listings generated from the YouTube Data API is fairly straightforward and relied on the fact that the paginated results from the API tended to remain the same over short periods of time depending on the token used to identify which page of a result set was being accessed.

This is where the concept of using an offset came from - if the results for the next page were relatively stable then you can fetch the next page, use the first part of it to fill the current page, then store the number of items used to fill the current page as the offset that is used to discard items from the next page when it is opened, because those items have already been displayed.

This was added towards the end of last year I believe, but since that time the way that the My Subscriptions listings were generated has changed, and as the output filtering methods have changed, the methods used to generate specific non-Data API listings also needs to be updated to suit the use of the fillers.

Additionally if the contents of the next page were not stable, then the use of an integer offset will no longer be valid for returning consistent results. This has become very apparent for the Recommended videos listing (another non-Data API based listing eg. https://github.com/anxdpanic/plugin.video.youtube/issues/1107#issuecomment-2741618718, https://github.com/anxdpanic/plugin.video.youtube/issues/1107#issuecomment-2739630320), which is why the filtering and filler methods need to be updated to work together.

The contents of the next page of a My Subscriptions listing may also no longer be stable - the My Subscription results were previously being fetched using a sort order based on the name of the associated subscribed channels, with individual videos then sorted based on upload date.

This meant that if you have a number of channels all updating on the same day then videos with channel names that sort higher alphabetically will always show up in the listings ahead of videos uploaded at the same time but with channel names that sort lower alphabetically. Related problem could occur if you had a channel update occur between generating a listing and navigating to the next page. 

To resolve this the subscriptions are now being fetched (and/or cached content updated) based on the channels unread content and/or when the new or total item count of the channel is different to when it was last fetched.

**Comment by ColonelPanic-42** [comment]
Thank you for the very thorough explanation @MoojMidge, that does clear a few things up (I had noticed there seemed to be different kinds of filtering which I understand now are the input vs output filtering). What you're saying re: the offset on its own isn't suitable makes sense, I've mainly been looking from the perspective of My Subscriptions, but I can imagine things are much more volatile in Recommended, and what you said re: sort order on My Subscriptions does shed some light on a weird and subtle behaviour I'd noticed a couple of times where items seemed to shift in order slightly usually a day or so between refreshes.

Based on your post it sounds like you've managed to solve this issue anyway, or at least worked towards it, so I will leave it alone for now and wait for your update 😁 these last few days spelunking through the code has been a good learning experience at least, and I've got a newfound appreciation for the work that's gone into it (and continues to go into it).

**Comment by MoojMidge** [comment]
Test version with updated filtering and pagination logic (https://github.com/MoojMidge/plugin.video.youtube/commit/4ea254f0887c28bd9b7b6a531d11514a7b2a52f7):
https://github.com/MoojMidge/plugin.video.youtube/archive/refs/heads/master.zip

**Comment by ColonelPanic-42** [comment]
> Test version with updated filtering and pagination logic ([MoojMidge@4ea254f](https://github.com/MoojMidge/plugin.video.youtube/commit/4ea254f0887c28bd9b7b6a531d11514a7b2a52f7)): https://github.com/MoojMidge/plugin.video.youtube/archive/refs/heads/master.zip

Initial testing looks promising 🥳 I'll update my main box with this and see how it goes over the next few days.

The only _potential_ bugbear I can think of is what impact that collection of "seen video ID's" has over time (at least in the short term). Haven't had a chance to confirm yet, but given the logic I imagine the following likely to happen:
* Load up e.g. "page" 2 in Kodi (which might be pages 3-5 from the YT feed) - video ID's A, B, C, D, E get consumed from page 5 and passed on to the query params for "page" 3 to exclude
* Loading "page" 3 immediately works as expected: A, B, C, D and E are filtered out as they were already consumed in the last "page", and Kodi displays from F onwards
* Tomorrow, refresh "page" 3. New items from page 1 push down the results, causing A, B, C, D, E to appear half way down page 5's results instead of at the top
* "Page" 3 now shows videos starting from e.g. -W, -X, -Y, -Z, _skips_ A, B, C, D, E, and jumps straight to F

IMO this is a much more bearable issue than dropping videos altogether as the current implementation does, but it's possibly something to be aware of (at the very least, to be prepared for some user to get confused by it). Could maybe be reduced by pushing the `exclude` list into the cache with a TTL of e.g. 1 day so calls to that page after that no longer exclude those videos, but IDK yet how worthwhile it would be to put in that effort. Otherwise we start talking about different pagination logic between feeds (My Subscriptions vs Recommended) which is a whole other can of worms that's best avoided if possible 😬

**Comment by MoojMidge** [comment]
> Haven't had a chance to confirm yet, but given the logic I imagine the following likely to happen:
> ...
> * Tomorrow, refresh "page" 3. New items from page 1 push down the results, causing A, B, C, D, E to appear half way down page 5's results instead of at the top
> * "Page" 3 now shows videos starting from e.g. -W, -X, -Y, -Z, _skips_ A, B, C, D, E, and jumps straight to F

Not sure, too lazy to think about it, but it will probably depend on how you access/refresh page 3.

A manual refresh will always clear the list of skipped videos, which can result in duplicates if skipping around in page history, but should never result in videos being skipped completely, which could happen previously when the offset parameter was still being used.

The filling of filtered results is a back fill i.e. will fill the current page using the next pages and then hide them on the next page so that nothing is repeated on the next pages as part of a sequential navigation sequence.

Refreshing a page or jumping to an arbitrary page would act to sever the link between the current page and the previous pages as it not a sequential navigation sequence, and the TTL of the list of back filled videos is intended to be valid only for the last sequential navigation action.

Trying to cache it would add a lot of complexity that I don't want to handle, due to how Kodi itself manages window history navigation.

**Comment by ColonelPanic-42** [comment]
Yeah I don't think it's a major issue, certainly not like losing videos altogether. TBH I wouldn't worry too much about trying to resolve the behaviour I mentioned unless it becomes more of an issue, it was just something to raise and keep in mind more than anything.

I haven't had a huge chance to test things out this week but the times I have been able to it seems to be working well 😁 haven't noticed any skipped videos missing between pages 🙌 although I do have to remember to get the "unofficial" build when I get the chance (recent versions keep using the massive tile view in Embuary and I want it to stick with the list view dammit 😂) I am looking forward to seeing the new listing techniques you're working on too

**Comment by benyamin-codez** [comment]
@MoojMidge - Looking good...! 👍

**Comment by MoojMidge** [comment]
As part of the listing offset no longer being used, additional methods for filling and filtering listings have now also been implemented.

It should mean that most listings now will have sequential page numbers, unless output filtering occurs, in which case the number of filtered items is now also shown.

Not well tested but you can try and see if you notice any issues prior to the next beta/stable release: https://github.com/MoojMidge/plugin.video.youtube/archive/refs/heads/master.zip
